File: app/src/main/java/code/fortomorrow/riceapp/HomeActivity.java
package code.fortomorrow.riceapp;

import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.ActionBarDrawerToggle;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.core.view.GravityCompat;
import androidx.drawerlayout.widget.DrawerLayout;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.firebase.ui.database.FirebaseRecyclerAdapter;
import com.firebase.ui.database.FirebaseRecyclerOptions;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.google.android.material.navigation.NavigationView;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.squareup.picasso.Picasso;

import code.fortomorrow.riceapp.Prevalent.Prevalent;
import code.fortomorrow.riceapp.ViewHolder.ProductViewHolder;
import code.fortomorrow.riceapp.utils.SpacesItemDecoration;
import de.hdodenhof.circleimageview.CircleImageView;
import io.paperdb.Paper;

public class HomeActivity extends AppCompatActivity implements NavigationView.OnNavigationItemSelectedListener {
    DrawerLayout drawerLayout;
    private DatabaseReference ProductsRef;
    private RecyclerView recyclerView;
    RecyclerView.LayoutManager layoutManager;
    private ImageView cart;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_home);
        ProductsRef = FirebaseDatabase.getInstance().getReference().child("Products");




        Paper.init(this);

        recyclerView = findViewById(R.id.recyler_menu);
        recyclerView.setHasFixedSize(true);
        //recyclerView.setHasFixedSize(true);
        layoutManager = new LinearLayoutManager(this);
        //recyclerView.addItemDecoration(new SpacesItemDecoration(10));
        recyclerView.setLayoutManager(layoutManager);

//        cart = findViewById(R.id.cart);
//        cart.setOnClickListener(new View.OnClickListener() {
//            @Override
//            public void onClick(View view) {
//                startActivity(new Intent(getApplicationContext(),CartActivity.class));
//            }
//        });

        Toolbar toolbar = findViewById(R.id.toolbar);
        toolbar.setTitle("Home");
        setSupportActionBar(toolbar);
        drawerLayout = findViewById(R.id.drawerLayout);
        NavigationView navigationView = findViewById(R.id.nav_view);
        ActionBarDrawerToggle toggle= new ActionBarDrawerToggle(this,drawerLayout,toolbar,R.string.navigation_drawer_open,R.string.navigation_drawer_close);
        drawerLayout.addDrawerListener(toggle);
        toggle.syncState();

        navigationView.setNavigationItemSelectedListener(this);
        FloatingActionButton fab = findViewById(R.id.nav_cart);
        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(HomeActivity.this,CartActivity.class);
                startActivity(intent);
            }
        });
        View headerView = navigationView.getHeaderView(0);
        TextView userNameTextView = headerView.findViewById(R.id.user_profile_name);
        CircleImageView profileImageView = headerView.findViewById(R.id.user_profile_image);

        userNameTextView.setText(Prevalent.currentOnlineUser.getName());
        Picasso.get().load(Prevalent.currentOnlineUser.getImage()).placeholder(R.drawable.profile).into(profileImageView);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            toggle.getDrawerArrowDrawable().setColor(getColor(R.color.colorPrimary));
        } else {
            toggle.getDrawerArrowDrawable().setColor(getResources().getColor(R.color.colorPrimary));
        }

    }

    @Override
    public boolean onNavigationItemSelected(@NonNull MenuItem menuItem) {
        int id = menuItem.getItemId();

        if (id == R.id.nav_cart)
        {
            Intent intent = new Intent(HomeActivity.this,CartActivity.class);
            startActivity(intent);
        }
        else if (id == R.id.nav_search)
        {
            Intent intent = new Intent(HomeActivity.this,SearchProductsActivity.class);
            startActivity(intent);
        }
//        else if (id == R.id.nav_categories)
//        {
//            startActivity(new Intent(getApplicationContext(),CategoryItemActivity.class));
//        }
        else if (id == R.id.nav_settings)
        {
            Intent intent = new Intent(HomeActivity.this, SettinsActivity.class);
            startActivity(intent);
        }
        else if (id == R.id.nav_logout)
        {
            Paper.book().destroy();

            Intent intent = new Intent(HomeActivity.this, LoginActivity.class);
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
            startActivity(intent);
            finish();
        }
        else if(id == R.id.about){
            Intent i = new Intent(getApplicationContext(),Aboutpage.class);
            startActivity(i);
        }

        DrawerLayout drawer = (DrawerLayout) findViewById(R.id.drawerLayout);
        drawer.closeDrawer(GravityCompat.START);
        return true;
    }

    @Override
    protected void onStart() {
        super.onStart();
        FirebaseRecyclerOptions<Products> options =
                new FirebaseRecyclerOptions.Builder<Products>()
                        .setQuery(ProductsRef, Products.class)
                        .build();


        FirebaseRecyclerAdapter<Products, ProductViewHolder> adapter =
                new FirebaseRecyclerAdapter<Products, ProductViewHolder>(options) {
                    @Override
                    protected void onBindViewHolder(@NonNull ProductViewHolder holder, int position, @NonNull final Products model)
                    {
                        holder.txtProductName.setText(model.getPname());
                        //holder.txtProductDescription.setText(model.getDescription());
                        holder.txtProductPrice.setText(model.getPrice() + " Taka");
                        Picasso.get().load(model.getImage()).into(holder.imageView);

                        holder.itemView.setOnClickListener(new View.OnClickListener() {
                            @Override
                            public void onClick(View v) {
                                Intent intent = new Intent(HomeActivity.this,ProductDetailsActivity.class);
                                intent.putExtra("pid",model.getPid());
                                startActivity(intent);
                            }
                        });
                    }

                    @NonNull
                    @Override
                    public ProductViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType)
                    {
                        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.products_items_layout, parent, false);
                        ProductViewHolder holder = new ProductViewHolder(view);
                        return holder;
                    }
                };
        recyclerView.setAdapter(adapter);
        adapter.startListening();
    }

    @Override
    public void onBackPressed() {
        if(drawerLayout.isDrawerOpen(GravityCompat.START)){
            drawerLayout.closeDrawer(GravityCompat.START);
        }
        else{
            super.onBackPressed();
        }
        super.onBackPressed();
    }
}

File: app/src/main/java/code/fortomorrow/happyshopping/view/HomeActivity.kt
package code.fortomorrow.happyshopping.view

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.core.view.GravityCompat
import androidx.drawerlayout.widget.DrawerLayout
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.StaggeredGridLayoutManager
import code.fortomorrow.happyshopping.R
import code.fortomorrow.happyshopping.prevalent.Prevalent
import code.fortomorrow.happyshopping.view_holder.ProductViewHolder
import com.firebase.ui.database.FirebaseRecyclerAdapter
import com.firebase.ui.database.FirebaseRecyclerOptions
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.google.android.material.navigation.NavigationView
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.FirebaseDatabase
import com.squareup.picasso.Picasso
import de.hdodenhof.circleimageview.CircleImageView
import io.paperdb.Paper

class HomeActivity : AppCompatActivity(), NavigationView.OnNavigationItemSelectedListener {
    private lateinit  var drawerLayout: DrawerLayout
    private var ProductsRef: DatabaseReference? = null
    private lateinit var recyclerView: RecyclerView
    var layoutManager: RecyclerView.LayoutManager? = null
    private val cart: ImageView? = null
    private lateinit var userNameTextView: TextView
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_home)
        ProductsRef = FirebaseDatabase.getInstance().reference.child("Products")


        Paper.init(this)

        recyclerView = findViewById(R.id.recyler_menu)
        recyclerView.setHasFixedSize(true)
        //recyclerView.setHasFixedSize(true);
        layoutManager = GridLayoutManager(this, 2)
        //        recyclerView.addItemDecoration(new SpacesItemDecoration(10));
        val staggeredGridLayoutManager = StaggeredGridLayoutManager(2, LinearLayoutManager.VERTICAL)


        recyclerView.setLayoutManager(staggeredGridLayoutManager)

        //        cart = findViewById(R.id.cart);
//        cart.setOnClickListener(new View.OnClickListener() {
//            @Override
//            public void onClick(View view) {
//                startActivity(new Intent(getApplicationContext(),CartActivity.class));
//            }
//        });
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        toolbar.title = "Home"
        setSupportActionBar(toolbar)
        drawerLayout = findViewById(R.id.drawerLayout)
        val navigationView = findViewById<NavigationView>(R.id.nav_view)
        val toggle = ActionBarDrawerToggle(
            this,
            drawerLayout,
            toolbar,
            R.string.navigation_drawer_open,
            R.string.navigation_drawer_close
        )
        drawerLayout.addDrawerListener(toggle)
        toggle.syncState()

        navigationView.setNavigationItemSelectedListener(this)
        val fab = findViewById<FloatingActionButton>(R.id.nav_cart)
        fab.setOnClickListener {
            val intent = Intent(this@HomeActivity, CartActivity::class.java)
            startActivity(intent)
        }
        val headerView = navigationView.getHeaderView(0)
        userNameTextView = headerView.findViewById(R.id.user_profile_name)
        val profileImageView = headerView.findViewById<CircleImageView>(R.id.user_profile_image)

        userNameTextView.setText(Prevalent.currentOnlineUser!!.name)
        Picasso.get().load(Prevalent.currentOnlineUser!!.image).placeholder(R.drawable.profile)
            .into(profileImageView)
        toggle.drawerArrowDrawable.color = getColor(R.color.colorPrimary)
    }

    override fun onNavigationItemSelected(menuItem: MenuItem): Boolean {
        val id = menuItem.itemId

        if (id == R.id.nav_cart) {
            val intent = Intent(this@HomeActivity, CartActivity::class.java)
            startActivity(intent)
        } else if (id == R.id.nav_search) {
            val intent = Intent(this@HomeActivity, SearchProductsActivity::class.java)
            startActivity(intent)
        } else if (id == R.id.nav_settings) {
            val intent = Intent(this@HomeActivity, SettinsActivity::class.java)
            startActivity(intent)
        } else if (id == R.id.nav_logout) {
            Paper.book().destroy()

            val intent = Intent(this@HomeActivity, LoginActivity::class.java)
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK)
            startActivity(intent)
            finish()
        } else if (id == R.id.myordersss) {
            val intent = Intent(this@HomeActivity, MyOrdersActivity::class.java)
            startActivity(intent)
        } else if (id == R.id.about) {
            val i = Intent(applicationContext, Aboutpage::class.java)
            startActivity(i)
        }

        val drawer = findViewById<View>(R.id.drawerLayout) as DrawerLayout
        drawer.closeDrawer(GravityCompat.START)
        return true
    }

    override fun onStart() {
        super.onStart()
        val options =
            FirebaseRecyclerOptions.Builder<Products>()
                .setQuery(ProductsRef!!, Products::class.java)
                .build()


        val adapter: FirebaseRecyclerAdapter<Products, ProductViewHolder> =
            object : FirebaseRecyclerAdapter<Products, ProductViewHolder>(options) {
                override fun onBindViewHolder(
                    holder: ProductViewHolder,
                    position: Int,
                    model: Products
                ) {
                    holder.txtProductName.text = model.pname
                    //holder.txtProductDescription.setText(model.getDescription());
                    holder.txtProductPrice.text = model.price + " Taka"
                    Picasso.get().load(model.image).into(holder.imageView)

                    holder.itemView.setOnClickListener {
                        val intent = Intent(this@HomeActivity, ProductDetailsActivity::class.java)
                        intent.putExtra("pid", model.pid)
                        startActivity(intent)
                    }
                }

                override fun onCreateViewHolder(
                    parent: ViewGroup,
                    viewType: Int
                ): ProductViewHolder {
                    val view = LayoutInflater.from(parent.context)
                        .inflate(R.layout.home_products_layout, parent, false)
                    return ProductViewHolder(view)
                }
            }
        recyclerView!!.adapter = adapter
        adapter.startListening()
    }

    override fun onBackPressed() {
        if (drawerLayout!!.isDrawerOpen(GravityCompat.START)) {
            drawerLayout!!.closeDrawer(GravityCompat.START)
        } else {
            super.onBackPressed()
        }
        super.onBackPressed()
    }
}
